### ISSUE TITLE
DistributedSampler shuffle option doesn't work as expected

### ISSUE THREAD

**Issue by nwoodall** [pytorch/pytorch#136588]
### 🐛 Describe the bug

When doing multi-GPU processing and using DistributedSampler, the shuffle=True setting results in some files in the dataset not being used on any of the GPU's. I found this when using the training script for evaluation and not turning off the shuffle. The expected number of output files written being lower than expected. I finally traced this to the fact that some of the dataset files were processed twice, and therefore those outputs were overridden, and some were not processed at all. Setting shuffle=False fixes the problem. This is an acceptable work around for evaluation, but it could have unintended consequences for training. 

Shuffle to me means that you are sampling without replacement. The behavior that I see is sampling without replacement for each GPU, but not across the GPU. That is, it is possible for different GPU to end up processing the same file.

Easy way to reproduce is to create a dataset equal to the number of GPU and printout which files each GPU is processing.

Pretty standard prepare DataLoader, only change is to allow different configurations for single vs. multiple GPUs. This code doesn't have the problem for evaluation because number of epochs for evaluation is always 1.

def prepare_dataloader(args, worldSize, dataset: Dataset, rng):
    if args.n_epochs == 1:
        shuffle_data = False # shuffling is not exclusive between GPU's which causes issues at evaluation
    else:
        shuffle_data = True
    if worldSize == 1:
        outLoader = DataLoader(
            dataset,
            batch_size=args.batch_size,
            num_workers=args.num_workers,
            pin_memory=True, 
            generator=rng,
            shuffle=True,
        )
    else:
        outLoader = DataLoader(
            dataset,
            batch_size=args.batch_size,
            num_workers=args.num_workers,
            pin_memory=True, 
            shuffle=False,
            sampler=DistributedSampler(dataset,num_replicas=worldSize,shuffle = shuffle_data)
        )
    return outLoader

### Versions

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.4.1
[pip3] torchaudio==2.4.1
[pip3] torchmetrics==1.4.2
[pip3] torchvision==0.19.1
[conda] libopenvino-pytorch-frontend 2024.4.0             hf9b8971_0    conda-forge
[conda] numpy                     1.26.4          py311h7125741_0    conda-forge
[conda] pytorch                   2.4.1                  py3.11_0    pytorch
[conda] torchaudio                2.4.1                 py311_cpu    pytorch
[conda] torchmetrics              1.4.2              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.19.1                py311_cpu    pytorch

**Comment by nwoodall** [comment]
This is a sample of the debug output where my dataset equals 8 unique files and the evaluation only processed 5 of the 8. The files processed by multiple GPUs are in bold.

[GPU: cuda:0] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:1] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:2] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:3] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:4] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:5] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:6] Epoch: 0 | B_SZ: 1 | Steps: 1
[GPU: cuda:7] Epoch: 0 | B_SZ: 1 | Steps: 1
[01:42<00:12, 10.50it/s]/tt/ssd0/neil/images/Output/**Out_1070_3.jpg**
last batch cuda:2
[01:42<00:05, 18.73it/s]/tt/ssd0/neil/images/Output/Out_1070_2.jpg
last batch cuda:4
[01:43<00:03, 25.41it/s]/tt/ssd0/neil/images/Output/Out_1070_1.jpg
last batch cuda:0
[01:45<00:01, 26.51it/s]/tt/ssd0/neil/images/Output/Out_1070_MFR2.jpg
last batch cuda:6
[01:46<00:00, 33.08it/s]/tt/ssd0/neil/images/Output/**Out_1070_3.jpg**
last batch cuda:3
[01:46<00:00, 33.99it/s]/tt/ssd0/neil/images/Output/**Out_1070_3.jpg**
last batch cuda:1
[01:46, 33.93it/s]/tt/ssd0/neil/images/Output/**Out_1070_MFR.jpg**
last batch cuda:5
[01:47<00:00, 36.59it/s]/tt/ssd0/neil/images/Output/**Out_1070_MFR.jpg**
last batch cuda:7
                                                                                            
(base) truecut@ubuntu:~/neil/pyproj$ cd /tt/ssd0/neil/images/Output
(base) truecut@ubuntu:/tt/ssd0/neil/images/Output$ ls
Out_1070_1.jpg  Out_1070_2.jpg  Out_1070_3.jpg  Out_1070_MFR2.jpg  Out_1070_MFR.jpg

**Comment by andrewkho** [comment]
@mikaylagawarecki thanks for this report! Could you share your dataset, dataloader, and sampler set up that created the test output?
